compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/SingleAbstractMethodLowering.kt

```diff
@@ -0,0 +1,231 @@
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
+import org.jetbrains.kotlin.load.java.sam.SamConstructorDescriptor
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.util.OperatorNameConventions
+
+internal val singleAbstractMethodPhase = makeIrFilePhase(
+    ::SingleAbstractMethodLowering,
+    name = "SingleAbstractMethod",
+    description = "Replace SAM conversions with instances of interface-implementing classes"
+)
+
+class SingleAbstractMethodLowering(val context: CommonBackendContext) : ClassLoweringPass {
+    override fun lower(irClass: IrClass) {
+        // TODO use `IrType`s; unfortunately, they are currently not comparable.
+        //      (Can't use class symbols here, we want type parameters as well.)
+        val cachedImplementations = mutableMapOf<Pair<KotlinType, KotlinType>, IrClass>()
+        val localImplementations = mutableListOf<IrClass>()
+        irClass.transformChildrenVoid(object : IrElementTransformerVoidWithContext() {
+            override fun visitCall(expression: IrCall): IrExpression {
+                if (expression.symbol.owner.descriptor !is SamConstructorDescriptor)
+                    return super.visitCall(expression)
+                // Direct construction of an interface:
+                //    Runnable { print("Hello, World!") }
+                return createInstance(expression.type, expression.getValueArgument(0)!!.transform(this, null))
+            }
+
+            override fun visitTypeOperator(expression: IrTypeOperatorCall): IrExpression {
+                if (expression.operator != IrTypeOperator.SAM_CONVERSION)
+                    return super.visitTypeOperator(expression)
+                // Implicit conversion to an interface:
+                //    // FILE: 1.java
+                //    class A { public static void f(Runnable r) { r.run() } }
+                //    // FILE: 2.kt
+                //    A.f { print("Hello, World!") }
+                return createInstance(expression.typeOperand, expression.argument.transform(this, null))
+            }
+
+            // Replace a SAM-conversion of a callable object with an instance of an appropriate implementation.
+            private fun createInstance(superType: IrType, invokable: IrExpression): IrExpression {
+                // Running in the class context, so empty scope means this is a field/anonymous initializer.
+                val scopeOwnerSymbol = currentScope?.scope?.scopeOwnerSymbol ?: irClass.symbol
+                return context.createIrBuilder(scopeOwnerSymbol).irBlock(invokable, null, superType) {
+                    // Not an exhaustive check, but enough to cover the most common cases.
+                    if (invokable is IrFunctionReference) {
+                        // Runnable(::function)
+                        createFunctionProxyInstance(superType, invokable)
```

Comment:
Please add bytecode test with number of generatedf class check

Reply:
Turns out the non-IR backend doesn't actually do that for references, only for lambdas. Should I remove this branch, or mark the test as IGNORE_BACKEND: JVM?

Reply:
@pyos Yes, mark as ignored for current backend